squash all ydb-support commit history

datbeohbbh opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

## Pull request type
Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?
Issue Number:

## What is the new behavior?

## Other information

datbeohbbh committed:
squash all ydb-support commit history

(temp fix) ignore secure connection test

- use `ydb` as hostname

fix go vet

fix CI

update ydb-go-sdk to v3.52.1

fix `GetTables`

asmyasnikov commented on .gitea/workflows/test-ydb.yml:
```diff
@@ -0,0 +1,61 @@
+    name: test ydb
+    runs-on: ubuntu-latest
+    steps:
+      # - name: cache go path
```
Is this really need in upstream?

datbeohbbh replied:
no, but all CI files in upstream have that comment part, I'm not sure if in future XORM maintainer will use it or not

asmyasnikov commented on Makefile:
```diff
@@ -47,6 +47,15 @@ TEST_DAMENG_HOST ?= dameng:5236
 TEST_DAMENG_USERNAME ?= SYSDBA
 TEST_DAMENG_PASSWORD ?= SYSDBA
 
+TEST_YDB_SCHEME ?= grpc
```
Maybe single parameter like `TEST_YDB_CONNECTION_STRING` instead 8 parameters?

asmyasnikov replied:
`TEST_YDB_CONNECTION_STRING=grpc://localhost:2136/local?qo_query_bind=table_path_prefix(/local/xorm/test),declare,numeric&go_fake_tx=scan,scheme,scripting`
or
`TEST_YDB_DSN=grpc://localhost:2136/local?qo_query_bind=table_path_prefix(/local/xorm/test),declare,numeric&go_fake_tx=scan,scheme,scripting`

datbeohbbh replied:
I will combine these parameters into one:
```
TEST_YDB_DBNAME ?= local
TEST_YDB_TABLE_PATH_PREFIX ?= /local/xorm/test
TEST_YDB_QUERY_BIND ?= table_path_prefix($(TEST_YDB_TABLE_PATH_PREFIX)),declare,numeric
TEST_YDB_FAKE_TX ?= scan,scheme,scripting
```
But will keep `scheme` and `host` for secure connection test
```
TEST_YDB_SCHEME ?= grpc
TEST_YDB_HOST ?= ydb:2136
```

asmyasnikov replied:
for secure connection you can change connection string to
```
TEST_YDB_CONNECTION_STRING=grpcs://localhost:2135/local?qo_query_bind=table_path_prefix(/local/xorm/test),declare,numeric&go_fake_tx=scan,scheme,scripting
```

asmyasnikov replied:
You have only one usage of envs TEST_YDB_SCHEME && TEST_YDB_HOST
I think you dont make difficult if you can make easy instead

asmyasnikov commented on dialects/ydb.go:
```diff
@@ -0,0 +1,1189 @@
+	switch v := t; v {
+	case schemas.Bool, schemas.Boolean:
+		yqlType = yql_Bool
+		return
```
replace to `return yql_Bool`?

asmyasnikov replied:
In all cases

asmyasnikov commented on dialects/ydb.go:
```diff
@@ -0,0 +1,1189 @@
+func yqlToSQLType(yqlType string) (sqlType schemas.SQLType) {
+	switch yqlType {
+	case yql_Bool:
+		sqlType = schemas.SQLType{Name: schemas.Bool, DefaultLength: 0, DefaultLength2: 0}
```
replace to `return  schemas.SQLType{Name: schemas.Bool, DefaultLength: 0, DefaultLength2: 0}`

asmyasnikov replied:
in all cases

asmyasnikov commented on dialects/ydb.go:
```diff
@@ -0,0 +1,1189 @@
+	return
+}
+
+func removeOptional(s string) string {
```
See `types.IsOptional` for check and get wrapped type

datbeohbbh replied:
To remove function `removeOptional` I think, SDK should be updated so that `func (r *rows) ColumnTypeDatabaseTypeName` and `func (c *conn) GetColumnType` only return wrapped type. What do you think about it?

datbeohbbh committed:
update ydb-go-sdk version to v3.52.2

datbeohbbh committed:
(refactor) add `TEST_YDB_CONNECTION_STRING`

(refactor) fix return in `toYQLDataType`

(refactor) fix returns in `yqlToSQLType`

datbeohbbh commented on dialects/time.go:
```diff
@@ -41,6 +45,9 @@ func FormatColumnTime(dialect Dialect, dbLocation *time.Location, col *schemas.C
 		return t.Format(layout), nil
 	case schemas.DateTime, schemas.TimeStamp:
+		if dialect != nil && dialect.URI().DBType == schemas.YDB {
+			return t, nil
```
suppose I have a struct `Series`:
```go
type Series struct {
	SeriesID    []byte    `xorm:"pk 'series_id'"`
	ReleaseDate time.Time `xorm:"'release_date' TIMESTAMP"`
}
```
when I insert value:
```go
_, err = engine.Insert(&Series{SeriesID: ID, ReleaseDate: time.Now()})
```
value of field `ReleaseDate` will be convert to interface through method `Value2Interface` and the converted value is the actual value to be passed to sdk:
https://github.com/ydb-platform/xorm/blob/066cbc34473ae958b8a89cd7eda48401a09660ad/internal/statements/values.go#L87-L91

as you can see, function `FormatColumnTime` will be called and if not returns raw time, a value of type `string` will be returned. Type `string` is not compatible and will cause errors.

datbeohbbh committed:
(fix) remove hacks in `session_insert` `(*Session) insertStruct`

datbeohbbh committed:
(fix) remove hacks for condition with custom type

datbeohbbh committed:
(fix) fix convert `null` value

datbeohbbh committed:
(fix) remove ad-hoc case for YDB

datbeohbbh committed:
(fix) remove ad-hoc cases for YDB

- remove ad-hoc cases in `sync.go`, `engine.go`. To remove, I decided
  in `dialects/ydb.go` method `CreateTableSQL` will only generate YQL to
  create table not YQL for create indexes. So, to create indexes for
  table, client should use `(*Session).CreateIndexes(...)`.

datbeohbbh committed:
(fix) remove ad-hoc case for YDB

- add method `GetDB()` for `core.Tx`, that method is used to obtain
  internal db.
